tlv/stream_test.go

```diff
@@ -0,0 +1,408 @@
+				tlv.MakePrimitiveRecord(1, &o.u16),
+				tlv.MakePrimitiveRecord(2, &o.u32),
+				tlv.MakePrimitiveRecord(3, &o.u64),
+				tlv.MakePrimitiveRecord(4, &o.buf),
+			)
+			return o, s
+		},
+		expObj: &testObj{
+			u16: 2,
+			u32: 3,
+			u64: 7,
+			buf: bytes.Repeat([]byte{0x05}, 3),
+		},
+	},
+}
+
+func TestTLV(t *testing.T) {
+	for _, test := range tlvTests {
+		t.Run(test.name, func(t *testing.T) {
+			testTLV(t, test)
+		})
+	}
+}
+
+func testTLV(t *testing.T, test tlvTestCase) {
+	_, s1 := test.encGen()
+
+	var b bytes.Buffer
+	err := s1.Encode(&b)
+	if err != nil {
+		t.Fatalf("unable to encode tlv stream: %v", err)
+	}
+
+	obj2, s2 := test.decGen()
+	err = s2.Decode(bytes.NewReader(b.Bytes()))
+	if err != nil {
+		t.Fatalf("unable to decode tlv stream: %v", err)
+	}
+
+	if !reflect.DeepEqual(test.expObj, obj2) {
+		t.Fatalf("mismatch object, want: %v, got: %v",
+			test.expObj, obj2)
+	}
+}
+
+type thing struct {
+	known uint32
+
+	field1   uint32
+	field9   uint32
+	field10  uint32
+	field100 uint32
+	field101 []byte
+
+	tlvStream *tlv.Stream
+}
+
+type CreateSessionTLV struct {
+	BlobType     blob.Type
+	MaxUpdates   uint16
+	RewardBase   uint32
+	RewardRate   uint32
+	SweepFeeRate lnwallet.SatPerKWeight
+
+	tlvStream *tlv.Stream
+}
+
+func NewCreateSessionTLV() *CreateSessionTLV {
+	m := &CreateSessionTLV{}
+	m.tlvStream = tlv.NewStream(
+		wtwire.WriteElement, wtwire.ReadElement,
+		tlv.MakeStaticRecord(0, &m.BlobType, 2),
+		tlv.MakePrimitiveRecord(1, &m.MaxUpdates),
+		tlv.MakePrimitiveRecord(2, &m.RewardBase),
+		tlv.MakePrimitiveRecord(3, &m.RewardRate),
+		tlv.MakeStaticRecord(4, &m.SweepFeeRate, 8),
+	)
+
+	return m
+}
+
+func (c *CreateSessionTLV) Encode(w io.Writer) error {
+	return c.tlvStream.Encode(w)
+}
+
+func (c *CreateSessionTLV) Decode(r io.Reader) error {
+	return c.tlvStream.Decode(r)
+}
+
+func newThing() *thing {
+	t := new(thing)
+
+	t.tlvStream = tlv.NewStream(
+		wtwire.WriteElement, wtwire.ReadElement,
+		[]tlv.Record{
+			tlv.MakePrimitiveRecord(1, &t.field1),
+			tlv.MakePrimitiveRecord(9, &t.field9),
+			tlv.MakePrimitiveRecord(10, &t.field10),
+			tlv.MakePrimitiveRecord(100, &t.field100),
+			tlv.MakePrimitiveRecord(101, &t.field101),
+		}...,
+	)
+
+	return t
+}
+
+func (t *thing) Encode(w io.Writer) error {
+	err := wtwire.WriteElement(w, t.known)
+	if err != nil {
+		return err
+	}
+
+	return t.tlvStream.Encode(w)
+}
+
+func (t *thing) EncodeNorm(w io.Writer) error {
+	return wtwire.WriteElements(w,
+		t.known,
+		t.field1,
+		t.field9,
+		t.field10,
+		t.field100,
+		t.field101,
+	)
+}
+
+func (t *thing) DecodeNorm(r io.Reader) error {
+	return wtwire.ReadElements(r,
+		&t.known,
+		&t.field1,
+		&t.field9,
+		&t.field10,
+		&t.field100,
+		&t.field101,
+	)
+}
+func (t *thing) Decode(r io.Reader) error {
+	err := wtwire.ReadElement(r, &t.known, 0)
+	if err != nil {
+		return err
+	}
+
+	return t.tlvStream.Decode(r)
+}
+
+func BenchmarkEncodeTLV(t *testing.B) {
```

Comment:
Test description would be good here to explain the different cases.

Reply:
added